Basic support for committing point cloud tile diffs

olsen232 opened the pull request:
<img src="https://media2.giphy.com/media/TijL9TieqPfLq/giphy.gif"/>

For each tile diff, commits it by
- copying the tile to the local LFS cache
- writing a pointer-file to the ODB
- updating the worktree-index so the diff disappears

Also makes sure that only tiles/*.laz are committable -
- QGIS likes to create random files next to tiles, that we don't want to commit

https://github.com/koordinates/kart/issues/565

rcoup commented on kart/lfs_util.py:
```diff
@@ -49,6 +78,26 @@ def pointer_file_to_json(pointer_file_bytes, result=None):
 
 
+def dict_to_pointer_file_bytes(pointer_dict):
+    def sort_key(key_value):
+        key, value = key_value
+        if key == "version":
+            return ""
+        return key
+
+    blob = bytearray()
+    for key, value in sorted(pointer_dict.items(), key=sort_key):
+        # TODO - LFS doesn't support our fancy pointer files yet. Hopefully fix this in LFS.
+        if key not in ("version", "oid", "size"):
+            continue
+        blob += key.encode("utf8")
+        blob += b" "
+        blob += str(value).encode("utf8")
+        blob += b"\n"
+
+    return blob
```
```suggestion
    blob = bytearray()
    for key, value in sorted(pointer_dict.items(), key=lambda i: (i[0] != "version", i)):
        # TODO - LFS doesn't support our fancy pointer files yet. Hopefully fix this in LFS.
        if key not in ("version", "oid", "size"):
            continue
        blob += f"{key} {value}\n".encode("utf8")

    return blob
```

rcoup replied:
> `if key not in ("version", "oid", "size"):`

This ignores all the `ext-*` ones, but I guess for now it's ok to ignore them? Nothing else will be writing Kart repos

olsen232 replied:
Yeah - this is a temporary workaround anyway. Plenty of time to fix it if we start writing extensions into pointer files

rcoup commented on kart/lfs_util.py:
```diff
@@ -61,6 +110,8 @@ def get_hash_from_pointer_file(pointer_file_bytes):
 def get_local_path_from_lfs_hash(repo, lfs_hash):
     """Given a sha256 LFS hash, finds where the object would be stored in the local LFS cache."""
+    if lfs_hash.startswith("sha256:"):
+        lfs_hash = lfs_hash[len("sha256:") :]
```
```suggestion
        lfs_hash = lfs_hash[len("sha256:"):]
```

Or:

```suggestion
        lfs_hash = lfs_hash[7:]  # len("sha256:")
```

olsen232 replied:
Went with the second. Black doesn't like the first one

rcoup reviewed (approved):
looks good 👍

olsen232 commented on kart/lfs_util.py:
```diff
@@ -22,16 +22,45 @@ def get_hash_and_size_of_file(path):
+    """
+    Given a path to a file, calculates and returns its SHA256 hash and length in bytes,
+    while copying it to the given destination.
+    """
```
Yeah to be honest I haven't profiled anything so this is probably a premature optimisation. Changed it to 1MB

olsen232 commented:
The diffs (as they are currently) are visible in test_point_cloud, here they are:
```
--- auckland:tile:auckland_1_1.copc.laz
+++ auckland:tile:auckland_1_1.copc.laz
-                                      oid = sha256:1d9934b50ebd057d893281813fda8deffb0ad03b3c354ec1c5d7557134c40be1
+                                      oid = sha256:dafd2ed5671190433ca1e7cea364a94d9e00c11f0a7b3927ce93554df5b1cd5c
-                                     size = 23570
+                                     size = 68665
--- auckland:tile:auckland_3_3.copc.laz
-                                     name = auckland_3_3.copc.laz
-                                      oid = sha256:522ef2ff7f66b51516021cde1fa7b9f301acde6713772958d6f1303fdac40c25
-                                     size = 1334
+++ auckland:tile:auckland_4_4.copc.laz
+                                     name = auckland_4_4.copc.laz
+                                      oid = sha256:522ef2ff7f66b51516021cde1fa7b9f301acde6713772958d6f1303fdac40c25
+                                     size = 1334
```

These diffs are somewhat limited -
- you can tell which tiles have been inserted, updated, deleted
- you can even find renames if you look at the hashes, but the code doesn't do it for you yet
- you *can't* tell anything else about the tiles from the diffs
  - this is partly because I haven't yet written the code to use PDAL to check what's in the files that the user is trying to add
  - and partly because I am currently not storing any tile metadata in pointer files, due it not being LFS compatible and I haven't yet implemented a workaround
  
More work still to be done of course
